src/tree/tree_model.cc

```diff
@@ -92,26 +93,50 @@ class TreeGenerator {
   virtual std::string SplitNode(RegTree const& tree, int32_t nid, uint32_t depth) {
     auto const split_index = tree[nid].SplitIndex();
     std::string result;
+    auto is_categorical = tree.GetSplitTypes()[nid] == FeatureType::kCategorical;
     if (split_index < fmap_.Size()) {
       switch (fmap_.TypeOf(split_index)) {
-        case FeatureMap::kIndicator: {
+      case FeatureMap::kCategorical: {
+        auto name = split_index < fmap_.Size()
+                        ? fmap_.Name(split_index)
+                        : "feat" + std::to_string(split_index);
```

Comment:
Since `name` is only needed for the error message, can you put this inside `if (!is_categorical)` block? Inside, use `LOG(FATAL)` to throw error

Reply:
Also, how come we are not writing anything to `result` for `case FeatureMap::kCategorical:`, like all the other cases?

Reply:
I kept the possibility that users don't provide a feature map with categorical type.  Categorical is a different split type, all the others are just variations of numerical splits.  So no matter what the feature map says it has to go through a dedicated print procedure.

I can restrict it and ensure a correct feature map is provided.

Reply:
I remove this logic, now users either not provide a feature map or it has to be consistent with the training data.